File: app/models/social_network.rb
class SocialNetwork < ActiveRecord::Base
  attr_accessible :name,
                  :scale,
                  :translation_x,
                  :translation_y,
                  :node_ids,
                  :family_ids,
                  :role_ids,
                  :user_id,
                  :description
  belongs_to :user
  has_many :nodes, dependent: :destroy
  has_many :node_attributes, through: :nodes
  has_many :families, dependent: :destroy
  has_many :roles, dependent: :destroy

  def uri(base = prefix)
    @uri ||= RDF::URI("#{base}/social_networks/#{id}")
  end

  # TODO: fix this is SO ugly :(
  def self.vocabulary
    url_helpers = Rails.application.routes.url_helpers
    if Rails.env.production?
      url_helpers.vocabulary_url(host: 'social-graph-editor.heroku.com')
    else
      url_helpers.vocabulary_url(host: 'sn2.dcc.uchile.cl')
    end
  end

  def prefix
    "http://sn.dcc.uchile.cl/2013"
  end

  def join!(imported, equivalences)
    equivalences.each_pair do |imported_family_id, original_family_id|
      imported_family = Family.find(imported_family_id)
      if original_family_id.nil? or original_family_id.empty?
        imported_family.update_attribute(:social_network, self)
      else
        original_family = Family.find(original_family_id)
        imported_family.nodes.each do |node|
          unless node.families.include?(original_family)
            node.families << original_family 
            node.save!
          end
        end
        imported_family.destroy
      end
    end

    imported.nodes.each do |node|
      node.update_attribute(:social_network, self)
    end

    imported.roles.each do |role|
      role.update_attribute(:social_network, self)
    end

    imported.reload.destroy
  end

end

File: app/assets/javascripts/models/social_network.js
App.SocialNetwork = DS.Model.extend({
  // attributes
  name: DS.attr('string'),
  description: DS.attr('string', { defaultValue: "" }),
  scale: DS.attr('number', { defaultValue: 1 }),
  translation_x: DS.attr('number', { defaultValue: 0 }),
  translation_y: DS.attr('number', { defaultValue: 0 }),

  // relations
  nodes: DS.hasMany('App.Node'),
  families: DS.hasMany('App.Family'),
  roles: DS.hasMany('App.Role'),
  user: DS.belongsTo('App.User'),

  translation: function() {
    return this.get('translation_x') + ", " + this.get('translation_y');
  }.property('translation_x', 'translation_y'),

  actorFamilies: function() {
    return this.get('families').toArray().filter(function (family) {
      return (family.get("kind") === "Actor");
    })
  }.property("families.@each.kind", "families.@each.name"),

  relationFamilies: function() {
    return this.get('families').toArray().filter(function (family) {
      return (family.get("kind") === "Relation");
    })
  }.property("families"),

  currentMode: function() {
    return "Hand"; // Other values "Actor", "Relation", "Role"
  }.property(),

  selectedFamily: function() {
    return null;
  }.property(),

  temporal: function() {
    return false; 
  }.property(),

  isNotTemporal: function() {
    return !this.get('temporal');
  }.property('temporal'),

});
